lib/api/objectCopy.js

```diff
@@ -0,0 +1,304 @@
+                        dataLocator = [];
+                    } else {
+                        // To provide for backwards compatibility before
+                        // md-model-version 2, need to handle cases where
+                        // objMD.location is just a string
+                        dataLocator = Array.isArray(sourceObjMD.location) ?
+                        sourceObjMD.location : [{ key: sourceObjMD.location }];
+                    }
+
+                    if (sourceObjMD['x-amz-server-side-encryption']) {
+                        for (let i = 0; i < dataLocator.length; i++) {
+                            dataLocator[i].masterKeyId = sourceObjMD
+                            ['x-amz-server-side-encryption-aws-kms-key-id'];
+                            dataLocator[i].algorithm =
+                                sourceObjMD['x-amz-server-side-encryption'];
+                        }
+                    }
+                    return next(null, storeMetadataParams, dataLocator);
+                });
+        },
+        function checkDestAuth(storeMetadataParams, dataLocator, next) {
+            return services.metadataValidateAuthorization(valPutParams,
+                (err, destBucketMD, destObjMD) => {
+                    if (err) {
+                        log.debug('error validating put part of request',
+                        { error: err });
+                        return next(err);
+                    }
+                    const flag = destBucketMD.hasDeletedFlag()
+                        || destBucketMD.hasTransientFlag();
+                    if (flag) {
+                        log.trace('deleted flag or transient flag ' +
+                        'on destination bucket', { flag });
+                        return next(errors.NoSuchBucket);
+                    }
+                    return next(null, storeMetadataParams,
+                        dataLocator, destBucketMD, destObjMD);
+                });
+        },
+        function goGetData(storeMetadataParams, dataLocator, destBucketMD,
+            destObjMD, next) {
+            const serverSideEncryption = destBucketMD.getServerSideEncryption();
+
+            // skip if source and dest the same or 0 byte object
+            // still send along serverSideEncryption info so algo
+            // and masterKeyId stored properly in metadata
+            if (sourceIsDestination || dataLocator.length === 0) {
+                return next(null, storeMetadataParams, dataLocator, destObjMD,
+                    serverSideEncryption);
+            }
+             // dataLocator is an array.  need to get and put all parts
+             // For now, copy 1 part at a time. Could increase the second
+             // argument here to increase the number of parts
+             // copied at once.
+            return async.mapLimit(dataLocator, 1,
+                // eslint-disable-next-line prefer-arrow-callback
+                function copyPart(part, cb) {
```

Comment:
I think it's worthwhile separating these functions into utility methods and call them. It would make it easier to follow the flow of logic. 
I would recommend 3 methods at least 

- copyPart
- encryptedPut
- unencryptedPut